src/leo_nlm_proto4_server.erl

```diff
@@ -0,0 +1,314 @@
+convert_lock({_, _, Owner, _, Offset, Length}, Excl) ->
+    End = case Length of
+              0 ->
+                  -1;
+              _ ->
+                  Offset + Length - 1
+          end,
+    #lock_record{
+       start = Offset,
+       till = End,
+       owner = Owner,
+       excl = Excl}.
+
+check_lock_ranges(_, []) ->
+    ok;
+check_lock_ranges(NewLock, [CurLock | Rest]) ->
+    Start = NewLock#lock_record.start,
+    Excl = NewLock#lock_record.excl,
+    Owner = NewLock#lock_record.owner,
+    CurStart = CurLock#lock_record.start,
+    CurEnd = CurLock#lock_record.till,
+    CurExcl = CurLock#lock_record.excl,
+    CurOwner = CurLock#lock_record.owner,
+    IsOverlap = case Start >= CurStart of
+                    true ->
+                        case CurEnd of
+                            -1 ->
+                                true;
+                            _ ->
+                                Start =< CurEnd
+                        end;
+                    false ->
+                        false
+                end,
+    case IsOverlap of
+        true ->
+            if 
+                Excl == false andalso CurExcl == false ->
+                    check_lock_ranges(NewLock, Rest);
+                Owner == CurOwner ->
+                    check_lock_ranges(NewLock, Rest);
+                true ->
+                    failed
+            end;
+        false ->
+            check_lock_ranges(NewLock, Rest)
+    end.
+
+modify_lock(#lock_record{start = CurStart}, Start, -1) when Start =< CurStart ->
+    [];
+modify_lock(#lock_record{start = CurStart} = Lock, Start, -1) when Start > CurStart ->
+    [Lock#lock_record{till = Start - 1}];
+modify_lock(#lock_record{start = CurStart, till = -1} = Lock, Start, End) ->
+    case End >= CurStart of
+        false ->
+            [];
+        true ->
+            Lock2 = Lock#lock_record{start = End + 1},
+            case Start =< CurStart of
+                true ->
+                    [Lock2];
+                false ->
+                    [Lock#lock_record{start = CurStart, till = Start - 1}, Lock2]
+            end
+    end;
+modify_lock(#lock_record{start = CurStart, till = CurEnd}, Start, End)
+  when Start =< CurStart, End >= CurEnd ->
+    [];
+modify_lock(#lock_record{start = CurStart, till = CurEnd} = Lock, Start, End)
+  when Start =< CurStart, End < CurEnd ->
+    [Lock#lock_record{start = End + 1}];
+modify_lock(#lock_record{start = CurStart, till = CurEnd} = Lock, Start, End)
+  when Start > CurStart, End >= CurEnd ->
+    [Lock#lock_record{till = Start - 1}];
+modify_lock(#lock_record{start = CurStart, till = CurEnd} = Lock, Start, End)
+  when Start > CurStart, End < CurEnd ->
+    ?debug("modify_lock", "Split ~p - ~p, ~p - ~p", [CurStart, Start-1, End + 1, CurEnd]),
+    Lock1 = Lock#lock_record{till = Start - 1},
+    Lock2 = Lock#lock_record{start = End + 1},
+    [Lock1, Lock2];
+modify_lock(Lock, _, _) ->
+    [Lock].
+
```

Comment:
Since convert_lock/2, check_lock_ranges/2 and modify_lock/3 are pure function,
We should have unit tests for those functions.